File: ios/Runner/BILMicSoundBridge.swift
import AVFoundation
import Flutter

/// Plays the two local microphone affordance sounds.
final class BILMicSoundBridge: NSObject, AVAudioPlayerDelegate {
  private let channel: FlutterMethodChannel
  private var player: AVAudioPlayer?
  private var playbackResult: FlutterResult?

  init(messenger: FlutterBinaryMessenger) {
    channel = FlutterMethodChannel(name: "bil/mic_sound", binaryMessenger: messenger)
    super.init()
    channel.setMethodCallHandler { [weak self] call, result in
      self?.handle(call, result: result)
    }
  }

  private func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    let name: String
    switch call.method {
    case "playOpen": name = "bil_mic_open_vibration"
    case "playEnd": name = "bil_mic_end_vibration"
    default:
      result(FlutterMethodNotImplemented)
      return
    }

    guard let url = Bundle.main.url(forResource: name, withExtension: "wav") else {
      result(FlutterError(code: "mic_sound_unavailable", message: nil, details: nil))
      return
    }
    do {
      player?.stop()
      finishPendingPlayback()
      let active = try AVAudioPlayer(contentsOf: url)
      active.delegate = self
      active.volume = 0.55
      active.prepareToPlay()
      player = active
      playbackResult = result
      guard active.play() else {
        player = nil
        finishPendingPlayback(
          FlutterError(code: "mic_sound_playback_rejected", message: nil, details: nil)
        )
        return
      }
    } catch {
      result(FlutterError(code: "mic_sound_failed", message: error.localizedDescription, details: nil))
    }
  }

  // Completing the MethodChannel call only after the 140 ms WAV finishes
  // makes Dart's `await BilMicSound.playOpen()` a real sequencing boundary:
  // the speech recognizer cannot open the microphone under the cue.
  func audioPlayerDidFinishPlaying(_ finishedPlayer: AVAudioPlayer, successfully flag: Bool) {
    guard player === finishedPlayer else { return }
    player = nil
    finishPendingPlayback(
      flag ? nil : FlutterError(code: "mic_sound_playback_failed", message: nil, details: nil)
    )
  }

  func audioPlayerDecodeErrorDidOccur(_ failedPlayer: AVAudioPlayer, error: Error?) {
    guard player === failedPlayer else { return }
    player = nil
    finishPendingPlayback(
      FlutterError(code: "mic_sound_failed", message: error?.localizedDescription, details: nil)
    )
  }

  private func finishPendingPlayback(_ error: FlutterError? = nil) {
    guard let callback = playbackResult else { return }
    playbackResult = nil
    if let error { callback(error) } else { callback(nil) }
  }

  func dispose() {
    player?.stop()
    player = nil
    finishPendingPlayback()
    channel.setMethodCallHandler(nil)
  }
}

File: android/app/src/main/kotlin/com/bilhealth/bodyintelligencelog/BILMicSoundBridge.kt
package com.bilhealth.bodyintelligencelog

import android.content.Context
import android.media.MediaPlayer
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

/** Plays the two local microphone affordance sounds. */
class BILMicSoundBridge(
    context: Context,
    messenger: BinaryMessenger,
) : MethodChannel.MethodCallHandler {
    private val methods = MethodChannel(messenger, "bil/mic_sound")
    private val appContext = context.applicationContext
    private var player: MediaPlayer? = null
    private var playbackResult: MethodChannel.Result? = null

    init {
        methods.setMethodCallHandler(this)
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        val resource = when (call.method) {
            "playOpen" -> com.bilhealth.bodyintelligencelog.R.raw.bil_mic_open_vibration
            "playEnd" -> com.bilhealth.bodyintelligencelog.R.raw.bil_mic_end_vibration
            else -> null
        }
        if (resource == null) {
            result.notImplemented()
            return
        }
        try {
            player?.release()
            finishPendingPlayback()
            player = MediaPlayer.create(appContext, resource)
            val active = player
            if (active == null) {
                result.error("mic_sound_unavailable", null, null)
                return
            }
            active.setVolume(0.55f, 0.55f)
            playbackResult = result
            active.setOnCompletionListener { completed ->
                if (player === completed) player = null
                completed.release()
                finishPendingPlayback()
            }
            active.setOnErrorListener { failed, _, _ ->
                if (player === failed) player = null
                failed.release()
                finishPendingPlayback("mic_sound_playback_failed", null)
                true
            }
            active.start()
        } catch (error: Exception) {
            player?.release()
            player = null
            if (playbackResult === result) {
                finishPendingPlayback("mic_sound_failed", error.message)
            } else {
                result.error("mic_sound_failed", error.message, null)
            }
        }
    }

    /**
     * The Dart caller awaits this reply before starting speech recognition.
     * Replying from MediaPlayer completion keeps the microphone closed for the
     * full 140 ms cue on every Android device rather than racing playback.
     */
    private fun finishPendingPlayback(code: String? = null, message: String? = null) {
        val callback = playbackResult ?: return
        playbackResult = null
        if (code == null) callback.success(null) else callback.error(code, message, null)
    }

    fun dispose() {
        player?.release()
        player = null
        finishPendingPlayback()
        methods.setMethodCallHandler(null)
    }
}
